### DISCUSSION TITLE
OG:Image able to be used with <img>?

### DISCUSSION THREAD

**Discussion by tazcarper** [vercel/community#1437] [Help]
I notice the examples all use the endpoint within meta tags. Is there anything stopping us from using it for normal HTML images?

I am trying to load about 30 images on a page that are all dynamically generated via my OG:Image endpoint. Works locally, but when I try on a build, I get this error and the majority of images fail to load.

`This invocation reused an I/O object (streams, request/response bodies, and others) created by a previous invocation, which is not permitted.`

Any clues why? 

Is there any way to return a base64 version from the response?

**Comment by someguy9** [comment]
Same issue here, can't pin down what triggers the error.

**Comment by utkarshkukreti** [comment]
I'm getting the same error when my endpoint is called twice very quickly (I have a page that embeds two OG images which the browser requests in parallel on page load). What makes this worse is that the failing requests sends an empty response with status 200 and this response is then cached forever, so those images never load until I redeploy the app.

For a simple reproduction, in the following code, change the number on the left of the dash to something unique, save this as an HTML file and load it in your browser. In my tests, only 1 or 2 images load successfully.

```html
<img src="https://og-examples.vercel.sh/api/vercel?0-0">
<img src="https://og-examples.vercel.sh/api/vercel?0-1">
<img src="https://og-examples.vercel.sh/api/vercel?0-2">
<img src="https://og-examples.vercel.sh/api/vercel?0-3">
<img src="https://og-examples.vercel.sh/api/vercel?0-4">
```

Right now this URL responds with status 200 and an empty response: https://og-examples.vercel.sh/api/vercel?0-2.

**Reply by utkarshkukreti** [reply]
The cause is likely in the `@vercel/og` package which seems to be caching some assets for reuse. The package source is not on GitHub so I checked the code here: https://www.npmjs.com/package/@vercel/og?activeTab=explore.

**Reply by utkarshkukreti** [reply]
As a temporary workaround, I've moved to rendering the images on serverless functions instead of edge by directly using satori and resvg. It's slower, but works.

```
const element = <div>...</div>
const svg = await satori(element, { width, height, fonts })
const resvg = new Resvg(svg)
const buffer = resvg.render().asPng()

response.setHeader('content-type', 'image/png')
response.setHeader('cache-control', 'public, max-age=86400')
response.end(buffer)
```

**Reply by someguy9** [reply]
You're amazing, going to test this out, I assume I'll need to import those packages separately

**Reply by utkarshkukreti** [reply]
Yes, the packages/imports are:

```
import { Resvg } from '@resvg/resvg-js'
import satori from 'satori'
```

Edit: and the `fonts` need to be specified manually. I don't need non-latin text or emoji support so I just added Source Sans Pro 400.

```
const fonts = [
  {
    name: 'Source Sans Pro',
    data: fs.readFileSync(
      path.resolve(
        process.cwd(),
        'src/source-sans-pro-all-400-normal.woff',
      ),
    ),
  },
]
```

**Reply by utkarshkukreti** [reply]
@amyegan I don't think this should be the selected answer, the bug still exists in @vercel/og and since that package isn't open source we can't post an issue anywhere.

**Reply by utkarshkukreti** [reply]
I can still reproduce the bug on the og-examples website. @amyegan

**Reply by amyegan** [reply]
Hi @utkarshkukreti. Are you talking about the [OG Image Playground](https://og-playground.vercel.app) or a different example site? If you can share a code example demonstrating exactly what you're trying to do and how it's failing, that would be very helpful 🙏

**Reply by utkarshkukreti** [reply]
@amyegan the problem we're facing is that when an edge function that renders an OG image is called multiple times in quick succession, it throws this error:

```
This invocation reused an I/O object (streams, request/response bodies, and others) created by a previous invocation, which is not permitted.
```

and the endpoint ends up returning an empty image.

This can be reproduced on the example images in @vercel/og README (https://www.npmjs.com/package/@vercel/og). If you save this as `something.html` and open it in your browser, it will make 5 requests to the same endpoint as the README in quick succession and usually 3 or 4 of them would fail:

```html
<img src="https://og-examples.vercel.sh/api/vercel?0-0">
<img src="https://og-examples.vercel.sh/api/vercel?0-1">
<img src="https://og-examples.vercel.sh/api/vercel?0-2">
<img src="https://og-examples.vercel.sh/api/vercel?0-3">
<img src="https://og-examples.vercel.sh/api/vercel?0-4">
```

This is what I see:

<img width="1308" alt="CleanShot 2023-03-08 at 16 58 10@2x" src="https://user-images.githubusercontent.com/217126/223702869-5f76e973-5ec3-4053-8b07-148a1066f1d6.png">

This seems to be a bug in @vercel/og only reproduced when an endpoint gets too many requests at the same time. There's no repo for @vercel/og or we would have opened an issue there.

**Reply by utkarshkukreti** [reply]
Just saw that there was a new release of @vercel/og (0.4.0) yesterday. I haven't had a chance to test it yet, maybe @tazcarper or @someguy9 can confirm if the bug still exists or not if they have tried.

**Reply by tomy0000000** [reply]
Just tested the latest `@vercel/og` (0.5.4), and the bug still persists.

**Reply by someguy9** [reply]
> Just tested the latest `@vercel/og` (0.5.4), and the bug still persists.

Same for me

**Comment by noskovvkirill** [comment]
Hi! Issue has been stale for a while, but the bug still exists. It's frustrating and selectively breaks in the production 🙁 Would be great to get the input from Vercel team!

This link **works**:
https://www.shelf.im/api/og/polaroid?url=https%253A%252F%252Fi.scdn.co%252Fimage%252Fab6761610000e5ebcdce7620dc940db079bf4952

While the other one **fails**:
https://www.shelf.im/api/og/polaroid?url=https%253A%252F%252Fi.scdn.co%252Fimage%252Fab6761610000e5eb5a00969a4698c3132a15fbb0

<img width="767" alt="Screenshot 2023-06-01 at 11 23 56" src="https://github.com/vercel/community/assets/50060130/0cda1fac-15bd-4e15-9846-7f8b8043c5a6">
<img width="809" alt="Screenshot 2023-06-01 at 11 24 18" src="https://github.com/vercel/community/assets/50060130/e4a98899-ce22-4400-a02a-bb0c92582e05">

**Comment by j485** [comment]
Can y'all please acknowledge / resolve this.

https://twitter.com/Jad_AE/status/1664657429907361795?s=20

**Comment by tomy0000000** [comment]
Discussion in the organization is not getting a lot of attention. For those looking for progress, head to vercel/satori#469, as the maintainer has assigned the issue to the v1 milestone.
